app/feature/travel-certificate/src/main/kotlin/com/hedvig/android/feature/travelcertificate/data/DownloadTravelCertificateUseCase.kt

```diff
@@ -0,0 +1,60 @@
+package com.hedvig.android.feature.travelcertificate.data
+
+import android.os.Environment
+import arrow.core.Either
+import arrow.core.raise.either
+import com.apollographql.apollo3.ApolloClient
+import com.hedvig.android.apollo.safeExecute
+import com.hedvig.android.apollo.toEither
+import com.hedvig.android.core.common.ErrorMessage
+import com.hedvig.android.core.common.await
+import com.hedvig.android.feature.travelcertificate.CoInsured
+import java.io.File
+import java.io.IOException
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import kotlinx.datetime.LocalDate
+import kotlinx.serialization.Serializable
+import octopus.TravelCertificateCreateMutation
+import octopus.type.TravelCertificateCreateCoInsured
+import octopus.type.TravelCertificateCreateInput
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okio.buffer
+import okio.sink
+import slimber.log.e
+
+private const val CERTIFICATE_NAME = "hedvigTravelCertificate.pdf"
+
+class DownloadTravelCertificateUseCase() {
+
+  suspend fun invoke(travelCertificateUri: TravelCertificateUrl): Either<ErrorMessage, TravelCertificateUri> =
+    withContext(Dispatchers.IO) {
+      either {
+        val request = Request.Builder()
+          .url(travelCertificateUri.uri)
+          .build()
+
+        val downloadedFile = File(
+          Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS),
+          CERTIFICATE_NAME,
+        )
+
+        try {
+          val response = OkHttpClient().newCall(request).await()
```

Comment:
Could we get the existing OkHttpClient instance which we have in our koin graph? So that we don't need to create a new client on each invocation of this function.

Reply:
I got some errors related to headers when I used that. This URL probably expects only "default" headers.

Reply:
Right, it probably didn't like that it was getting some extra Authorization header with a token that it did not recognize. Let's leave it for now then